scripts/foundry/contract-upgrades/2.1.0/README.md

```diff
@@ -0,0 +1,51 @@
+1. Upgrade ChallengeManager to v2.1.0
+2. Set the wasm module root to the new version
+3. Set the conditional wasm root and one step proof
+4. Upgrade RollupAdminLogic and RollupUserLogic contracts
+
+Note that contracts without code changes are not upgraded. It is normal to have some contracts still in the old version after the upgrade as they are equivalent to the new version. After the contract upgrade, you would need to schedule an ArbOS upgrade to ArbOS31 to enable the new features.
+
+## Requirements
+This upgrade only support upgrading from the following [nitro-contract release](https://github.com/OffchainLabs/nitro-contracts/releases):
+- Inbox: v1.1.0 - v2.1.0 inclusive
+- Outbox: v1.1.0 - v2.1.0 inclusive
+- SequencerInbox: v1.2.1 - v2.1.0 inclusive
+- Bridge: v1.1.0 - v2.1.0 inclusive
+- RollupProxy: v1.1.0 - v2.1.0 inclusive
+- RollupAdminLogic: v1.1.0 - v2.1.0 inclusive
+- RollupUserLogic: v1.1.0 - v2.1.0 inclusive
+- ChallengeManager: v1.2.1 - v1.3.0 inclusive
+
+Please refer to the top [README](../../README.md) `Check Version and Upgrade Path` on how to determine your current nitro contracts version.
+
+## Deployed instances
+
+TBD
+
+## How to use it
+
+1. Setup .env according to the example files, make sure you have everything correctly defined. The script will do some sanity checks but not everything can be checked.
+
+2. (Skip this step if you can use the deployed instances of action contract) 
+`DeployNitroContracts2Point1Point0UpgradeActionScript.s.sol` script deploys templates, and upgrade action itself. It can be executed in this directory like this:
+```bash
+forge script --sender $DEPLOYER --rpc-url $PARENT_CHAIN_RPC --broadcast --slow ./DeployNitroContracts2Point1Point0UpgradeAction.s.sol -vvv --verify --skip-simulation
+# use --account XXX / --private-key XXX / --interactive / --ledger to set the account to send the transaction from
+```
+As a result, all templates and upgrade action are deployed. Note the last deployed address - that's the upgrade action.
+
+3. `ExecuteNitroContracts2Point1Point0Upgrade.s.sol` script uses previously deployed upgrade action to execute the upgrade. It makes following assumptions - L1UpgradeExecutor is the rollup owner, and there is an EOA which has executor rights on the L1UpgradeExecutor. Proceed with upgrade using the owner account (the one with executor rights on L1UpgradeExecutor):
+```bash
+forge script --sender $EXECUTOR --rpc-url $PARENT_CHAIN_RPC --broadcast ./ExecuteNitroContracts2Point1Point0Upgrade.s.sol -vvv
+# use --account XXX / --private-key XXX / --interactive / --ledger to set the account to send the transaction from
+```
+If you have a multisig as executor, you can still run the above command without broadcasting to get the payload for the multisig transaction.
+
+4. That's it, upgrade has been performed. You can make sure it has successfully executed by checking wasm module root:
+```bash
+cast call --rpc-url $PARENT_CHAIN_RPC $ROLLUP "wasmModuleRoot()"
+```
```

Comment:
Can add some FAQ

Q. node error: unable to find validator machine directory for the on-chain WASM module root 
```
unable to find validator machine directory for the on-chain WASM module root err="stat /home/user/target/machines/0x260f5fa5c3176a856893642e149cf128b5a8de9f828afec8d11184415dd8dc69: no such file or directory"
```
A: upgrade to nitro-node v3.1.0+

Q: error generating node action: wasmroot doesn't match rollup
```
error acting as staker                   err="error advancing stake from node 0 (hash 0x0000000000000000000000000000000000000000000000000000000000000000): error generating node action: wasmroot doesn't match rollup : 0x260f5fa5c3176a856893642e149cf128b5a8de9f828afec8d11184415dd8dc69, valid: [0x8b104a2e80ac6165dc58b9048de12f301d70b02a0ab51396c22b4b4b802a16a4]"
```
A: make sure there are activity on the child chain and a new batch is posted after the upgrade

Reply:
https://github.com/OffchainLabs/orbit-actions/pull/16/commits/12778ca7c78a298b97eb9b41fb18fa7b540c0969